### PR TITLE
chore(i18n): ensure that translations can be added and replaced at runtime

### LINKED ISSUE #11721: No way to replace transalation at runtime
`add_translation()` does weird thing summing arrays

### PR DESCRIPTION
**Pull request by hypeJunction**
Adds missing tests

### CODE REVIEW

**Review by jdalsem — APPROVED** [review]

File: engine/classes/Elgg/I18n/Translator.php [hunk 1]
```diff
@@ -188,18 +188,16 @@ public function addTranslation($country_code, $language_array) {
...
-			if (!isset($this->translations[$country_code])) {
-				$this->translations[$country_code] = $language_array;
-			} else {
-				$this->translations[$country_code] = $language_array + $this->translations[$country_code];
-			}
+		if (!isset($this->translations[$country_code])) {
+			$this->translations[$country_code] = [];
 		}
 
+		$this->translations[$country_code] = array_merge($this->translations[$country_code], $language_array);
```
Comment (jeabakker) [inline engine/classes/Elgg/I18n/Translator.php]: `array_merge` is a heavy computational function and it gets heavier the larger the array is. That's why a long loooong time ago this was changed to a `+` (see https://github.com/Elgg/Elgg/commit/7049bad2caedaad3aa84de9ba1141943bcec9772#diff-f67f4259ce99534800557ed80ab9015a)

Comment (hypeJunction) [reply engine/classes/Elgg/I18n/Translator.php]: Very interesting. The difference in performance is considerable.
http://sandbox.onlinephpfunctions.com/code/78497c6594b2d8839064417e974117fbe792d7fb

File: engine/tests/phpunit/unit/Elgg/I18n/TranslatorUnitTest.php [hunk 2]
```diff
@@ -182,4 +182,31 @@ public function testCanDetectCurrentLanguageFromConfig() {
...
+		$this->assertEquals('Dummy', $this->translator->translate('__elgg_php_unit:test_key', [], 'new_lang_code'));
+		$this->assertEquals('__elgg_php_unit:test_key2', $this->translator->translate('__elgg_php_unit:test_key2', [], 'new_lang_code'));
+
+		$this->translator->addTranslation('en', [
+			'__elgg_php_unit:test_key' => 'Not So Dummy',
+			'__elgg_php_unit:test_key2' => 'Still Dummy',
+		]);
+
+		$this->translator->addTranslation('new_lang_code', [
+			'__elgg_php_unit:test_key' => 'Not So Dummy',
```
Comment (jdalsem) [inline engine/tests/phpunit/unit/Elgg/I18n/TranslatorUnitTest.php]: could you change the translated string to something different than that from the `en` translation? That way we are sure it is not coming from the `en` language when getting it.

**Comment by jdalsem** [comment]
i am still not sure what this PR is actually fixing... what usecase is causing the problems?

**Comment by hypeJunction** [comment]
Last time I tried to add translation at runtime it didn't work, probably for some other reason. I wanted to fix but after writing a test there was no problem. So it's just an additional test to have

File: engine/tests/phpunit/unit/Elgg/I18n/TranslatorUnitTest.php (hunk 2 above)
Comment (hypeJunction) [reply engine/tests/phpunit/unit/Elgg/I18n/TranslatorUnitTest.php]: Done

**Pull request merged**